Corrects the metadata of normal projections

amosbird opened the pull request:
### Changelog category (leave one):
- Backward Incompatible Change

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):

Corrects the metadata of normal projections so that projections with multi-column sorting keys are properly recognized. Builds on top of #90429.

### Upgrade notes

If you have normal (non-aggregate) projections that use expressions (e.g. `dictGet(...) AS value ORDER BY value`), their metadata was stored incorrectly on previous versions. After upgrading, run the following for each affected projection:

```sql
ALTER TABLE <table> CLEAR PROJECTION <projection_name>;
ALTER TABLE <table> MATERIALIZE PROJECTION <projection_name>;
```

This will re-create the projection with the correct metadata.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)


<!-- ch-version-info:start -->
### Version info
- Merged into: `26.3.1.518`
<!-- ch-version-info:end -->

GrigoryPervakov committed:
Use implicit key columns from projections

amosbird committed:
Fix metadata of normal projections

amosbird committed:
Fix test

amosbird commented:
@GrigoryPervakov The previous normal projections have incorrect information stored in their metadata. Now it's fixed. Could you help take a look?

GrigoryPervakov reviewed (commented):
LGTM, but I'd prefer to show it to someone with deeper knowledge in projections

KochetovNicolai reviewed (approved):
Ok, but we need to revrite this code to the new analyzer

amosbird committed:
Merge remote-tracking branch 'upstream/master' into cond_90429

amosbird committed:
Use analyzer

amosbird committed:
Merge remote-tracking branch 'upstream/master' into cond_90429

amosbird committed:
Merge remote-tracking branch 'upstream/master' into cond_90429

amosbird committed:
Expand alias in normal projection's order by

amosbird committed:
Strip ALIAS column definitions for Analyzer in projection metadata

Use ColumnsDescription(columns.getAll()) to present all columns as
Ordinary to the Analyzer. This prevents initializeTableExpressionData
from eagerly resolving ALIAS column expressions, which may fail when
session settings (e.g. allow_nonconst_timezone_arguments) are not
available in the DDL context (CREATE/ATTACH TABLE).

amosbird committed:
Fix projections in 03755_circular_dictionary test

dictGet is non-deterministic and cannot be used as a sorting key.
Convert normal projections to aggregate projections using GROUP BY
instead, which was the original intent of the test.

amosbird committed:
Fix positional GROUP BY in DatabaseReplicated mode

The Analyzer's replaceNodesWithPositionalArguments skips positional
argument resolution when query_kind != INITIAL_QUERY (e.g. SECONDARY_QUERY
in DatabaseReplicated). Our pre-analysis step using runOnlyResolve would
then fail validateAggregates because GROUP BY 1, 2 stays as literal
constants instead of being resolved to column references.

Fix by setting query_kind to INITIAL_QUERY on the copied context before
running the Analyzer passes.

amosbird committed:
Isolate Analyzer context from InterpreterSelectQuery to preserve access control

The setQueryKindInitial() call needed for positional GROUP BY resolution in
the Analyzer was leaking into InterpreterSelectQuery via shared mut_context.
In DatabaseReplicated mode, this altered ClientInfo fields (initial_user,
query_kind) causing privilege checks to be bypassed — e.g. a user without
SELECT privilege on a table referenced in a projection subquery could
successfully create the projection.

Fix: use a scoped analyzer_context with setQueryKindInitial() only for the
Analyzer pre-analysis step, and create a separate mut_context (without
setQueryKindInitial()) for InterpreterSelectQuery.

amosbird committed:
Revert FetchColumns for non-aggregate projections to fix access control

The FetchColumns processing stage causes InterpreterSelectQuery to return
early before analyzing subqueries, which bypasses privilege checks. This
caused test_replicated_access::test_initiator_user_in_ddl to fail because
a projection with a subquery referencing a restricted table would succeed
instead of being rejected.

Reverting to WithMergeableState (which master always used) ensures the
full query analysis runs, including subquery access control checks. The
actual fix for projection metadata (using Analyzer-based is_aggregate
detection) is independent of the processing stage.

amosbird committed:
Revert "Revert FetchColumns for non-aggregate projections to fix access control"

This reverts commit b3e48686b908216b5d7efe007356841c20aed7d4.

amosbird committed:
Fix scalar subquery cache leak in projection pre-analysis

The runOnlyResolve step uses addQueryTreePasses() without only_analyze=true,
causing the Analyzer to actually execute scalar subqueries and cache their
results in the global query context. When InterpreterSelectQuery runs
afterward, it finds the cached result and skips creating a new interpreter,
thus bypassing access control checks for subqueries in projection definitions.

Pass only_analyze=true so that:
1. Scalar subqueries use getSampleBlock() instead of actual execution
2. Results go to local cache only (not the shared query context)
3. Cache keys get '_analyze' suffix to prevent cross-phase collisions

amosbird committed:
Simplify: use single context for both Analyzer and InterpreterSelectQuery

setQueryKindInitial() only changes query_kind and fills OS/version info,
it does not alter initial_user or other access-control-sensitive fields.
Combined with only_analyze=true preventing scalar cache pollution, a single
mut_context can safely serve both the Analyzer pre-analysis and the
InterpreterSelectQuery phases.

amosbird committed:
Fix test_initiator_user_in_ddl: expect success after projection fix

Michicosun commented on tests/queries/0_stateless/03755_circular_dictionary.sql:
```diff
@@ -139,8 +139,8 @@ CREATE TABLE background_metrics
-            dictGet('background_dict', 'value', metric) AS value
-        ORDER BY value
+            min(dictGet('background_dict', 'value', metric))
+        GROUP BY metric
```
What do you think - will this kind of test be fixed after this patch?

```
CREATE TABLE background_metrics
(
    `metric` String,
    `x` Int64,
    PROJECTION values
    (
        SELECT
            metric,
            dictGet('background_dict', 'value', metric) as value
        ORDER BY value
    )
)
ENGINE = MergeTree
ORDER BY metric;
```

I suppose this should work, but right now `KeyDescription::getSortingKeyFromAST` tries to lookup alias (`value`) in the columns list of storage. 

---

Also, a lot of problems are shadowed by 

```
/// Only keep required columns
Block source_block = block;
for (const auto & column : required_columns)
    source_block.insert(block.getByName(column));
```

Here, it should be more correct to initialize the source block as an empty block initially and extend it with `required_columns`, `_row_exists`, and `_part_offset` columns after.

I tried that and found 2 problems:
1. Test above
2. Subcolumns in the sorting key of the projection are not properly materialized in this case, and because of that there is another test case:

```
create table test
(
    json JSON,
    projection prj (select json, json.a order by json.b::Int32)
) engine=MergeTree order by ();

insert into test (json) select '{"a" : 42, "b" : 42}';
```

amosbird replied:
Thanks for the analysis!

To clarify the design: **normal projections only store raw table columns**. All non-raw columns are reduced to the original columns they depend on:

- `a + b` → stores `a` and `b`
- `json.a` → stores `json`
- `dictGet('d', 'v', metric) AS value` → stores `metric`

The subcolumn handling (keeping parent column, not adding subcolumn to metadata) already follows this principle correctly.

The current bug — which this PR fixes — is that computed expression columns in SELECT were **incorrectly being stored as materialized columns** in the projection metadata. After this fix, the metadata columns only contain original table columns. At read time, the sorting key's expression DAG derives expressions from the stored original columns.

Regarding your two examples:

**1. `dictGet` in ORDER BY** — `dictGet` is non-deterministic (`isDeterministic() = false`), so it cannot be used in a normal projection's ORDER BY. This should error at definition time.

**2. Subcolumns in sorting key** (`ORDER BY json.b::Int32` with `json` in SELECT) — Since `json` is stored, `json.b` is derivable at read time. This should work once the sorting key is built against the projection's stored columns.

For future work: we can introduce a projection setting to enable **materialized expression storage**, where computed columns are stored physically and the optimizer matches queries against them (similar to aggregate projections). That would be a separate PR.

amosbird reviewed (commented):
(duplicate)
